### ISSUE TITLE
[YSQL] Add support for advisory_locks 

### ISSUE THREAD

**Issue by ddorian** [yugabyte/yugabyte-db#3642]
Jira Link: [DB-2071](https://yugabyte.atlassian.net/browse/DB-2071)

Advisory locks is not yet supported in Yugabyte (as of 2024.1). This issue tracks the support for Advisory locks. 

With the fix for #19974 (commit 809199f5), the use of advisory locks fails with _"advisory locks are not yet implemented"_ error. If the application doesn't strictly need advisory locks or doesn't want to fail the statement(as a temporary stop-gap), the GUC - **yb_silence_advisory_locks_not_supported_error** can be used as a stop-gap solution to avoid disruption in their application until they modify the application to remove usage of advisory locks (or till Advisory locks is natively supported in YB).

**Main tasks corresponding to the feature**

- [x] https://github.com/yugabyte/yugabyte-db/issues/25127
- [x] https://github.com/yugabyte/yugabyte-db/issues/24357
- [x] https://github.com/yugabyte/yugabyte-db/issues/24709
  - [x] https://github.com/yugabyte/yugabyte-db/issues/25195 
  - [x] https://github.com/yugabyte/yugabyte-db/issues/24711
  - [x] https://github.com/yugabyte/yugabyte-db/issues/25238 
  - [x] https://github.com/yugabyte/yugabyte-db/issues/24710
  - [x] https://github.com/yugabyte/yugabyte-db/issues/25565
- [x] https://github.com/yugabyte/yugabyte-db/issues/24713

**Fit and Finish**
- [x] https://github.com/yugabyte/yugabyte-db/issues/25802
- [x] https://github.com/yugabyte/yugabyte-db/issues/25753 
- [x] https://github.com/yugabyte/yugabyte-db/issues/25761
- [x] https://github.com/yugabyte/yugabyte-db/issues/25585
- [x] https://github.com/yugabyte/yugabyte-db/issues/25638
- [x] https://github.com/yugabyte/yugabyte-db/issues/26180
- [x] https://github.com/yugabyte/yugabyte-db/issues/26518
- [x] https://github.com/yugabyte/yugabyte-db/issues/26513
- [ ] https://github.com/yugabyte/yugabyte-db/issues/25646
- [ ] https://github.com/yugabyte/yugabyte-db/issues/25566
- [x] https://github.com/yugabyte/yugabyte-db/issues/25293

**Supportability tasks**

- [x] https://github.com/yugabyte/yugabyte-db/issues/25939
- [x] https://github.com/yugabyte/yugabyte-db/issues/26057
- [x] https://github.com/yugabyte/yugabyte-db/issues/24712
- [x] https://github.com/yugabyte/yugabyte-db/issues/26110

**Example**

script.py
```
def pg_advisory():
    import psycopg2
    conn1 = psycopg2.connect("dbname='postgres' user='postgres' host=127.0.0.1 password='postgres' port=5432")
    conn2 = psycopg2.connect("dbname='postgres' user='postgres' host=127.0.0.1 password='postgres' port=5432")
    cur1 = conn1.cursor()
    cur2 = conn2.cursor()
    cur1.execute("SELECT pg_try_advisory_lock(1);")
    print("session_1_lock", cur1.fetchone())
    cur1.execute("SELECT pg_try_advisory_xact_lock(17);")
    print("transaction_1_lock", cur1.fetchone())
    cur2.execute("SELECT pg_try_advisory_lock(1);")
    print("session_2_lock", cur2.fetchone())
    cur2.execute("SELECT pg_try_advisory_xact_lock(17);")
    print("transaction_2_lock", cur2.fetchone())
    exit()

pg_advisory()
```

Yugabyte output:

```
session_1_lock (True,)
transaction_1_lock (True,)
session_2_lock (True,)
transaction_2_lock (True,)
```
Postgresql output:
```
session_1_lock (True,)
transaction_1_lock (True,)
session_2_lock (False,)
transaction_2_lock (False,)
```

[DB-2071]: https://yugabyte.atlassian.net/browse/DB-2071?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by Amandeepsinghghai** [comment]
Hi,
Coming from the slack channel https://yugabyte-db.slack.com/archives/CG0KQF0GG/p1596795769135800. 

Summarizing the chat here incase the history is not maintained -

- Our application is split into microservices, each of which use a separate schema on the same YSQL database.
- We use [golang-migrate](https://github.com/golang-migrate/migrate) which generates an advisory lock based on the database name + schema name (not the whole database).
- We use K8S Init Containers to run our migrations. However, when our microservices are starting up in parallel, only the first one's migrations succeed, and then rest fail.

Do we need to perhaps write some custom code to use a separate Locks table (referring to golang-migrate's [cockroachdb implementation](https://github.com/golang-migrate/migrate/blob/master/database/cockroachdb/cockroachdb.go#L178) vs their [postgres](https://github.com/golang-migrate/migrate/blob/master/database/postgres/postgres.go#L169) implementation)? We're creating a workaround to prevent our microservices from starting up simultaneously, but that prevents us from having our microservices be independently deployable.

**Comment by ddorian** [comment]
Hi @Amandeepsinghghai 

Yes some custom code would be needed here. I think even the Mysql implementation should do in our case.
Can you try that ?

**Comment by Amandeepsinghghai** [comment]
Hey @ddorian, thanks for the reply. By MYSQL, I am guessing you mean [this](https://github.com/golang-migrate/migrate/blob/master/database/mysql/mysql.go). 

Isn't `"SELECT GET_LOCK(?, 10)"` **NOT** part of the Postgres standard or maybe I missing something here?

Also looping in my teammate cc @RabidFire

**Comment by ddorian** [comment]
@Amandeepsinghghai yes, `GET_LOCK` is not supported in PostgreSQL/YugabyteDB.

The CRDB workaround should work in our case too.

**Comment by Amandeepsinghghai** [comment]
Okay. Let me try that and see if it works.

**Comment by Amandeepsinghghai** [comment]
Okay. Let me see if that works.

**Comment by ddorian** [comment]
@Amandeepsinghghai that's because CRDB doesn't support user defined schemas.

While we do support them. So multiple schemas should work with us.

**Comment by jameshuynh** [comment]
Facing the same issue when provision postgres database using TF - https://registry.terraform.io/providers/cyrilgdn/postgresql/latest/docs/resources/postgresql_database

**Comment by yogendra** [comment]
@jameshuynh 
I would recommend silencing the advisory lock request errors with 
```
ALTER ROLE <DB_UserName> SET yb_silence_advisory_locks_not_supported_error=on;
```

Advisory locks are for avoiding concurrent DDLs. In tools like terraform you are already locking the concurrent script execution, so its safe to silence the errors. Additionally, I have seen that advisory locks being a session level construct have issues when executing on PG also in certain configurations.

Silencing advisory locks support error does not impact any DMLs.

**Comment by sameer-m-dev** [comment]
Hi YB team, are we planning on adding this feature anytime soon?

**Comment by sunsided** [comment]
Just stumbled upon this while setting up hopscotch (see [here](https://docs.hoppscotch.io/documentation/self-host/community-edition/install-and-build#running-migrations)).

**Comment by rthallamko3** [comment]
@sameer-m-dev , The feature is on our roadmap, we are designing the feature currently and are making progress towards adding the support.

**Comment by simon-mottram-cucumber** [comment]
Using Flyway to migrate hits this error as well. Will have to wait for fix

Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'flywayInitializer' defined in class path resource [org/springframework/boot/autoconfigure/flyway/FlywayAutoConfiguration$FlywayConfiguration.class]: Unable to acquire PostgreSQL advisory lock
------------------------------------------
SQL State  : 0A000
Error Code : 0
Message    : ERROR: advisory locks are not yet implemented
  Hint: See https://github.com/yugabyte/yugabyte-db/issues/3642. React with thumbs up to raise its priority

**Comment by yogendra** [comment]
You can use the suggestion in [earlier reply](https://github.com/yugabyte/yugabyte-db/issues/3642#issuecomment-2024495729) to circumvent this for now.

**Comment by rthallamko3** [comment]
Advisory locks is available in preview, starting from Yugabyte build 2.25.1.0-b180.

 The preview gflag `ysql_yb_enable_advisory_locks` needs to be enabled on both TServer and Master to try the feature.

**If you are using yugabyted, you need to enable the flags as shown below:**
 
`yugabyted start --master_flags="allowed_preview_flags_csv={ysql_yb_enable_advisory_locks},ysql_yb_enable_advisory_locks=true" --tserver_flags="allowed_preview_flags_csv={ysql_yb_enable_advisory_locks},ysql_yb_enable_advisory_locks=true" 
`

**Comment by vigkastHV** [comment]
> Advisory locks is available in preview, starting from Yugabyte build 2.25.1.0-b180.
> 
> The preview gflag `ysql_yb_enable_advisory_locks` needs to be enabled on both TServer and Master to try the feature.
> 
> **If you are using yugabyted, you need to enable the flags as shown below:**
> 
> `yugabyted start --master_flags="allowed_preview_flags_csv={ysql_yb_enable_advisory_locks},ysql_yb_enable_advisory_locks=true" --tserver_flags="allowed_preview_flags_csv={ysql_yb_enable_advisory_locks},ysql_yb_enable_advisory_locks=true" `


Hey @rthallamko3. I couldn't see this version in helm chart. Do I specifically need to deploy it from docker? Can you please suggest a way to how to proceed on this?

And is this available in  2.25.0.0-b489?

**Comment by rthallamko3** [comment]
@vigkastHV , The changes are not available in 2.25.0 version. The 2.25.1.0 version will be released in the month of March.

**Comment by vigkastHV** [comment]
> [@vigkastHV](https://github.com/vigkastHV) , The changes are not available in 2.25.0 version. The 2.25.1.0 version will be released in the month of March.

Hey @rthallamko3 . With the support of pg_adivsory_lock in version 2.25.1.0, Is there a support for distributed-locks as well in this release?

**Comment by rthallamko3** [comment]
@vigkastHV , Was your question about table level locks? That is tracked by https://github.com/yugabyte/yugabyte-db/issues/11571

**Comment by rthallamko3** [comment]
@jonathansp , We are preparing the official release of 2.25.1.0 version in early march. Will post an update here once it is available for folks.

**Comment by jonathansp** [comment]
@rthallamko3 how does the progress go so far?

**Comment by rthallamko3** [comment]
@jonathansp , I believe the release team is preparing to publish the release. Should happen this week or next week. Will keep you posted.

**Comment by jonathansp** [comment]
many thanks @rthallamko3!

**Comment by rthallamko3** [comment]
@jonathansp , Were you able to try out the feature, did you have any feedback for us. We are looking to see if any of the additional optimizations listed are necessary or folks are happy with the things in the preview.

**Comment by pruiz** [comment]
If additional feedback is welcomed, we are starting to move less demanding workloads over to yugabytedb, but not having advisory locks in LTS is slowing us quite a bit, as we need to rework some application internals in order to perform locking differently, etc.

**Comment by rthallamko3** [comment]
@pruiz , Advisory locks will be included in the upcoming 2025.1 stable release. The feature is currently in preview, you can test it out on the 2.25.1 preview build as mentioned in [above](https://github.com/yugabyte/yugabyte-db/issues/3642#issuecomment-2749031537). What exactly is your application scenario?

**Comment by rthallamko3** [comment]
With https://github.com/yugabyte/yugabyte-db/issues/25293, The Advisory locks feature is on by default in master branch and in upcoming 2025.1 stable release. Closing this master github as this concludes the main feature development effort.

The minor bug fixes that are tracked in the corresponding issues will be closed out in the 2025.1 timeframe.

The future enhancements items will be tracked in a new issue - https://github.com/yugabyte/yugabyte-db/issues/27131
